gadgets/trace_tcp/program.bpf.c

```diff
@@ -0,0 +1,406 @@
+		return true;
+
+	return false;
+}
+
+static __always_inline int enter_tcp_connect(struct pt_regs *ctx,
+					     struct sock *sk)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = pid_tgid;
+	__u64 uid_gid = bpf_get_current_uid_gid();
+	__u32 uid = uid_gid;
+	__u64 mntns_id;
+
+	mntns_id = gadget_get_mntns_id();
+
+	if (filter_event(sk, uid, pid, mntns_id))
+		return 0;
+
+	bpf_map_update_elem(&sockets, &tid, &sk, 0);
+	return 0;
+}
+
+static __always_inline int exit_tcp_connect(struct pt_regs *ctx, int ret,
+					    __u16 family)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = pid_tgid;
+	__u64 uid_gid = bpf_get_current_uid_gid();
+	struct tuple_key_t tuple = {};
+	struct pid_comm_t pid_comm = {};
+	struct sock **skpp;
+	struct sock *sk;
+	struct task_struct *task;
+
+	skpp = bpf_map_lookup_elem(&sockets, &tid);
+	if (!skpp)
+		return 0;
+
+	if (ret)
+		goto end;
+
+	sk = *skpp;
+
+	if (!fill_tuple(&tuple, sk, family))
+		goto end;
+
+	task = (struct task_struct *)bpf_get_current_task();
+
+	pid_comm.pid = pid;
+	pid_comm.uid_gid = uid_gid;
+	pid_comm.mntns_id = (u64)BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
+	bpf_get_current_comm(&pid_comm.comm, sizeof(pid_comm.comm));
+
+	bpf_map_update_elem(&tuplepid, &tuple, &pid_comm, 0);
+
+end:
+	bpf_map_delete_elem(&sockets, &tid);
+	return 0;
+}
+
+SEC("kprobe/tcp_v4_connect")
+int BPF_KPROBE(ig_tcp_v4_co_e, struct sock *sk)
+{
+	return enter_tcp_connect(ctx, sk);
+}
+
+SEC("kretprobe/tcp_v4_connect")
+int BPF_KRETPROBE(ig_tcp_v4_co_x, int ret)
+{
+	return exit_tcp_connect(ctx, ret, AF_INET);
+}
+
+SEC("kprobe/tcp_v6_connect")
+int BPF_KPROBE(ig_tcp_v6_co_e, struct sock *sk)
+{
+	return enter_tcp_connect(ctx, sk);
+}
+
+SEC("kretprobe/tcp_v6_connect")
+int BPF_KRETPROBE(ig_tcp_v6_co_x, int ret)
+{
+	return exit_tcp_connect(ctx, ret, AF_INET6);
+}
+
+SEC("kprobe/tcp_close")
+int BPF_KPROBE(ig_tcp_close, struct sock *sk)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u64 uid_gid = bpf_get_current_uid_gid();
+	__u32 uid = uid_gid;
+	struct tuple_key_t tuple = {};
+	struct event *event;
+	u16 family;
+	__u64 mntns_id;
+
+	mntns_id = gadget_get_mntns_id();
+
+	if (filter_event(sk, uid, pid, mntns_id))
+		return 0;
+
+	/*
+	 * Don't generate close events for connections that were never
+	 * established in the first place.
+	 */
+	u8 oldstate = BPF_CORE_READ(sk, __sk_common.skc_state);
+	if (oldstate == TCP_SYN_SENT || oldstate == TCP_SYN_RECV ||
+	    oldstate == TCP_NEW_SYN_RECV)
+		return 0;
+
+	family = BPF_CORE_READ(sk, __sk_common.skc_family);
+	if (!fill_tuple(&tuple, sk, family))
+		return 0;
+
+	event = gadget_reserve_buf(&events, sizeof(*event));
+	if (!event)
+		return 0;
+
+	fill_event(&tuple, event, pid, uid_gid, family, TCP_EVENT_TYPE_CLOSE,
+		   mntns_id);
+	bpf_get_current_comm(&event->task, sizeof(event->task));
+	event->timestamp = bpf_ktime_get_boot_ns();
```

Comment:
Move these two lines inside `fill_event`  to avoid repeating many times.

Reply:
That's a good idea! Though, I think only  `event->timestamp = bpf_ktime_get_boot_ns();` can be moved inside `fill_event` as `BPF_KPROBE(ig_tcp_state)` seems to set `event->task` in [L347](https://github.com/inspektor-gadget/inspektor-gadget/blob/claudia/convert-trace-tcp-gadget/gadgets/trace_tcp/program.bpf.c#L347) *not* to the current process:
```c
__builtin_memcpy(&event->task, p->comm, sizeof(event->task));
``` 
Am I misunderstanding?